File: backend/app/hubs.py
"""Indian logistics hub network — coordinates, corridor distances, toll density.

Verbatim port of `src/lib/hubs.ts` (data copied 1:1 so distances and corridor
profiles match the Next.js implementation exactly).
"""
from __future__ import annotations

import math
from dataclasses import dataclass


class UnknownHubError(KeyError):
    """Raised when a city is not part of the hub network."""

    def __init__(self, city: str) -> None:
        super().__init__(city)
        self.city = city

    def __str__(self) -> str:  # message must start with "Unknown hub" (API maps it to INVALID_HUB)
        return f"Unknown hub: {self.city}"


@dataclass(frozen=True)
class Hub:
    city: str
    state: str
    lat: float
    lng: float


HUBS: tuple[Hub, ...] = (
    Hub("Mumbai", "Maharashtra", 19.076, 72.8777),
    Hub("Pune", "Maharashtra", 18.5204, 73.8567),
    Hub("Nashik", "Maharashtra", 19.9975, 73.7898),
    Hub("Nagpur", "Maharashtra", 21.1458, 79.0882),
    Hub("Delhi", "Delhi NCR", 28.6139, 77.209),
    Hub("Gurugram", "Haryana", 28.4595, 77.0266),
    Hub("Jaipur", "Rajasthan", 26.9124, 75.7873),
    Hub("Ahmedabad", "Gujarat", 23.0225, 72.5714),
    Hub("Surat", "Gujarat", 21.1702, 72.8311),
    Hub("Ludhiana", "Punjab", 30.901, 75.8573),
    Hub("Kolkata", "West Bengal", 22.5726, 88.3639),
    Hub("Bhubaneswar", "Odisha", 20.2961, 85.8245),
    Hub("Raipur", "Chhattisgarh", 21.2514, 81.6296),
    Hub("Hyderabad", "Telangana", 17.385, 78.4867),
    Hub("Vijayawada", "Andhra Pradesh", 16.5062, 80.648),
    Hub("Bengaluru", "Karnataka", 12.9716, 77.5946),
    Hub("Chennai", "Tamil Nadu", 13.0827, 80.2707),
    Hub("Coimbatore", "Tamil Nadu", 11.0168, 76.9558),
    Hub("Kochi", "Kerala", 9.9312, 76.2673),
    Hub("Indore", "Madhya Pradesh", 22.7196, 75.8577),
    Hub("Kanpur", "Uttar Pradesh", 26.4499, 80.3319),
    Hub("Guwahati", "Assam", 26.1445, 91.7362),
)

_HUBS_BY_LOWER: dict[str, Hub] = {h.city.lower(): h for h in HUBS}


def hub(city: str) -> Hub:
    """Look up a hub by city name (case-insensitive). Raises UnknownHubError."""
    h = _HUBS_BY_LOWER.get(city.lower())
    if h is None:
        raise UnknownHubError(city)
    return h


def has_hub(city: str) -> bool:
    return city.lower() in _HUBS_BY_LOWER


def distance_km(a_lat: float, a_lng: float, b_lat: float, b_lng: float) -> int:
    """Great-circle distance in km with the 1.18 road factor (as in hubs.ts)."""
    r = 6371.0
    d_lat = math.radians(b_lat - a_lat)
    d_lng = math.radians(b_lng - a_lng)
    s = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(a_lat)) * math.cos(math.radians(b_lat)) * math.sin(d_lng / 2) ** 2
    )
    return round(2 * r * math.asin(math.sqrt(s)) * 1.18)


@dataclass(frozen=True)
class CorridorDef:
    from_city: str
    to_city: str
    toll_density: float  # 0..1
    congestion: float  # 0..1
    weather_risk: float  # 0..1 seasonal


CORRIDORS: tuple[CorridorDef, ...] = (
    CorridorDef("Mumbai", "Pune", 0.8, 0.65, 0.55),
    CorridorDef("Mumbai", "Delhi", 0.75, 0.6, 0.35),
    CorridorDef("Mumbai", "Bengaluru", 0.7, 0.55, 0.3),
    CorridorDef("Delhi", "Kolkata", 0.65, 0.5, 0.5),
    CorridorDef("Bengaluru", "Chennai", 0.7, 0.5, 0.4),
    CorridorDef("Delhi", "Jaipur", 0.6, 0.45, 0.25),
    CorridorDef("Hyderabad", "Bengaluru", 0.6, 0.4, 0.3),
    CorridorDef("Kolkata", "Chennai", 0.5, 0.45, 0.6),
    CorridorDef("Ahmedabad", "Mumbai", 0.65, 0.55, 0.3),
    CorridorDef("Nagpur", "Hyderabad", 0.5, 0.35, 0.35),
    CorridorDef("Delhi", "Ludhiana", 0.55, 0.5, 0.45),
    CorridorDef("Coimbatore", "Kochi", 0.4, 0.35, 0.5),
    CorridorDef("Kolkata", "Guwahati", 0.35, 0.3, 0.65),
    CorridorDef("Mumbai", "Kolkata", 0.6, 0.5, 0.45),
    CorridorDef("Chennai", "Kochi", 0.45, 0.35, 0.5),
)

File: frontend/src/lib/hubs.ts
/** Indian logistics hub network — coordinates, corridor distances, toll density. */

export interface Hub {
  city: string;
  state: string;
  lat: number;
  lng: number;
}

export const HUBS: Hub[] = [
  { city: "Mumbai", state: "Maharashtra", lat: 19.076, lng: 72.8777 },
  { city: "Pune", state: "Maharashtra", lat: 18.5204, lng: 73.8567 },
  { city: "Nashik", state: "Maharashtra", lat: 19.9975, lng: 73.7898 },
  { city: "Nagpur", state: "Maharashtra", lat: 21.1458, lng: 79.0882 },
  { city: "Delhi", state: "Delhi NCR", lat: 28.6139, lng: 77.209 },
  { city: "Gurugram", state: "Haryana", lat: 28.4595, lng: 77.0266 },
  { city: "Jaipur", state: "Rajasthan", lat: 26.9124, lng: 75.7873 },
  { city: "Ahmedabad", state: "Gujarat", lat: 23.0225, lng: 72.5714 },
  { city: "Surat", state: "Gujarat", lat: 21.1702, lng: 72.8311 },
  { city: "Ludhiana", state: "Punjab", lat: 30.901, lng: 75.8573 },
  { city: "Kolkata", state: "West Bengal", lat: 22.5726, lng: 88.3639 },
  { city: "Bhubaneswar", state: "Odisha", lat: 20.2961, lng: 85.8245 },
  { city: "Raipur", state: "Chhattisgarh", lat: 21.2514, lng: 81.6296 },
  { city: "Hyderabad", state: "Telangana", lat: 17.385, lng: 78.4867 },
  { city: "Vijayawada", state: "Andhra Pradesh", lat: 16.5062, lng: 80.648 },
  { city: "Bengaluru", state: "Karnataka", lat: 12.9716, lng: 77.5946 },
  { city: "Chennai", state: "Tamil Nadu", lat: 13.0827, lng: 80.2707 },
  { city: "Coimbatore", state: "Tamil Nadu", lat: 11.0168, lng: 76.9558 },
  { city: "Kochi", state: "Kerala", lat: 9.9312, lng: 76.2673 },
  { city: "Indore", state: "Madhya Pradesh", lat: 22.7196, lng: 75.8577 },
  { city: "Kanpur", state: "Uttar Pradesh", lat: 26.4499, lng: 80.3319 },
  { city: "Guwahati", state: "Assam", lat: 26.1445, lng: 91.7362 },
];

export function hub(city: string): Hub {
  const h = HUBS.find((x) => x.city.toLowerCase() === city.toLowerCase());
  if (!h) throw new Error(`Unknown hub: ${city}`);
  return h;
}

export function hasHub(city: string): boolean {
  return HUBS.some((x) => x.city.toLowerCase() === city.toLowerCase());
}

/** Great-circle distance in km. */
export function distanceKm(aLat: number, aLng: number, bLat: number, bLng: number): number {
  const R = 6371;
  const dLat = ((bLat - aLat) * Math.PI) / 180;
  const dLng = ((bLng - aLng) * Math.PI) / 180;
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((aLat * Math.PI) / 180) * Math.cos((bLat * Math.PI) / 180) * Math.sin(dLng / 2) ** 2;
  return Math.round(2 * R * Math.asin(Math.sqrt(s)) * 1.18); // 1.18 road factor
}

export interface CorridorDef {
  from: string;
  to: string;
  tollDensity: number; // 0..1
  congestion: number; // 0..1
  weatherRisk: number; // 0..1 seasonal
}

export const CORRIDORS: CorridorDef[] = [
  { from: "Mumbai", to: "Pune", tollDensity: 0.8, congestion: 0.65, weatherRisk: 0.55 },
  { from: "Mumbai", to: "Delhi", tollDensity: 0.75, congestion: 0.6, weatherRisk: 0.35 },
  { from: "Mumbai", to: "Bengaluru", tollDensity: 0.7, congestion: 0.55, weatherRisk: 0.3 },
  { from: "Delhi", to: "Kolkata", tollDensity: 0.65, congestion: 0.5, weatherRisk: 0.5 },
  { from: "Bengaluru", to: "Chennai", tollDensity: 0.7, congestion: 0.5, weatherRisk: 0.4 },
  { from: "Delhi", to: "Jaipur", tollDensity: 0.6, congestion: 0.45, weatherRisk: 0.25 },
  { from: "Hyderabad", to: "Bengaluru", tollDensity: 0.6, congestion: 0.4, weatherRisk: 0.3 },
  { from: "Kolkata", to: "Chennai", tollDensity: 0.5, congestion: 0.45, weatherRisk: 0.6 },
  { from: "Ahmedabad", to: "Mumbai", tollDensity: 0.65, congestion: 0.55, weatherRisk: 0.3 },
  { from: "Nagpur", to: "Hyderabad", tollDensity: 0.5, congestion: 0.35, weatherRisk: 0.35 },
  { from: "Delhi", to: "Ludhiana", tollDensity: 0.55, congestion: 0.5, weatherRisk: 0.45 },
  { from: "Coimbatore", to: "Kochi", tollDensity: 0.4, congestion: 0.35, weatherRisk: 0.5 },
  { from: "Kolkata", to: "Guwahati", tollDensity: 0.35, congestion: 0.3, weatherRisk: 0.65 },
  { from: "Mumbai", to: "Kolkata", tollDensity: 0.6, congestion: 0.5, weatherRisk: 0.45 },
  { from: "Chennai", to: "Kochi", tollDensity: 0.45, congestion: 0.35, weatherRisk: 0.5 },
];
